File: 518. Coin Change 2/518. Coin Change 2 top-down.java
import java.util.Arrays;

class Solution {
    private int n;
    private int[] coins;
    private int[][] memo;

    private int dp(int amount, int coin_index) {
        if (amount < 0 || coin_index >= n) {
            return 0;
        }
        if (amount == 0) {
            return 1;
        }
        if (memo[amount][coin_index] == -1) {
            int take = dp(amount - coins[coin_index], coin_index);
            int notake = dp(amount, coin_index + 1);
            memo[amount][coin_index] = take + notake;
        }
        return memo[amount][coin_index];
    }

    public int change(int amount, int[] coins) {
        this.coins = coins;
        this.n = coins.length;
        this.memo = new int[amount + 1][coins.length];
        for (int[] arr : memo)
            Arrays.fill(arr, -1);
        return dp(amount, 0);
    }
}

File: 518. Coin Change 2/518. Coin Change 2 top-down.py
from functools import lru_cache
from typing import List


class Solution:
    def change(self, amount: int, coins: List[int]) -> int:
        n = len(coins)

        @lru_cache(None)
        def dp(amount: int, coin_index: int) -> int:
            # base case
            if coin_index >= n or amount < 0:
                return 0
            if amount == 0:
                return 1

            # recurrence relation
            take = dp(amount-coins[coin_index], coin_index)
            notake = dp(amount, coin_index+1)

            return take + notake

        return dp(amount, 0)


s = Solution()
amount = 5
coins = [1, 2, 5]
print(s.change(amount, coins))
